server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -942,6 +947,30 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
     }
 
     private void indexValue(DocumentParserContext context, long timestamp) {
+        // DataStreamTimestampFieldMapper and TsidExtractingFieldMapper need to use timestamp value,
+        // so when this is true we store it in a well-known place
+        // instead of forcing them to iterate over all fields.
+        //
+        // DataStreamTimestampFieldMapper is present and enabled both
+        // in data streams and standalone indices in time_series mode
+        if (isDataStreamTimestampField && context.mappingLookup().isDataStreamTimestampFieldEnabled()) {
```

Comment:
This `isDataStreamTimestampFieldEnabled` check is unfortunate. The reason for it being here is that we can't fully trust `MapperBuilderContext#isDataStream`. In its current implementation `isDataStream` starts as `false` and then gets flipped to `true` if `DataStreamTimestampFieldMapper` is enabled in **current** mapping. This is broken for an edge case of a mapping overriding default `@timestamp` mapping.

Scenario:
1. Create a time series index with default mapping only
2. Update mapping with
```
{"@timestamp": {"type": "date"}}
```
For this mapping update `isDataStream` is `false` even though it was `true` previously.

One option to fix this is to pass existing mapping into `MappingParser#parse` and initialize `isDataStream` to a value based on it. I didn't want to mix this change in here.

Reply:
Thanks for this explanation. Let's make fixing `MapperBuilderContext#isDataStream()` a followup of this change. As part of that change the ` && context.mappingLookup().isDataStreamTimestampFieldEnabled()` part of the condition can be removed.